### PR TITLE
Add optional `--exclude=<name>` argument for multiple plugin commands

### PR DESCRIPTION
**Pull request by danielbachhuber**
Previously #149, #195

---

Added optional argument `--exclude` for `plugin deactivate --all` where plugin slugs can be specified as comma-separated.

Objective: Allow users to exclude specific plugins from deactivation.
E.g. `wp plugin deactivate --all --exclude=jetpack,akismet`

Also, the `--exclude` parameter will work only if `--all` is specified. I've added 2 additional tests to check this feature under `features/plugin-deactivate.feature`

-- @achyuthajoy

### CODE REVIEW

File: features/plugin-activate.feature [hunk 1]
```diff
@@ -106,3 +106,12 @@ Feature: Activate WordPress plugins
...
       """
+
+  Scenario: Adding --exclude with plugin activate --all should exclude the plugins specified via --exclude
+    When I try `wp plugin activate --all --exclude=hello`
+    Then STDOUT should be:
+      """
+      Plugin 'akismet' activated.
+      Success: Activated 1 of 1 plugins.
+      """
+    And the return code should be 0
```
Comment (swissspidy) [inline features/plugin-activate.feature:117]: Nit: Missing newline at the end of the file

File: features/plugin-deactivate.feature [hunk 2]
```diff
@@ -74,5 +74,23 @@ Feature: Deactivate WordPress plugins
...
+    And the return code should be 0
+
+  Scenario: Adding --exclude with plugin deactivate should throw an error unless --all given
+    When I try `wp plugin deactivate --exclude=hello`
+    Then the return code should be 1
+    And STDERR should be:
+      """
+      Error: Please specify one or more plugins, or use --all.
+      """
+    And STDOUT should be empty
```
Comment (swissspidy) [inline features/plugin-deactivate.feature:96]: Nit: Missing newline at the end of the file

File: features/plugin-delete.feature [hunk 3]
```diff
@@ -39,3 +39,11 @@ Feature: Delete WordPress plugins
...
       """
     And the return code should be 0
+
+  Scenario: Excluding a plugin from deletion when using --all switch
+    When I try `wp plugin delete --all --exclude=akismet,hello`
+    Then STDOUT should be:
+      """
+      Success: No plugins deleted.
+      """
+    And the return code should be 0
```
Comment (swissspidy) [inline features/plugin-delete.feature:49]: Nit: Missing newline at the end of the file

File: features/plugin-uninstall.feature [hunk 4]
```diff
@@ -71,3 +71,11 @@ Feature: Uninstall a WordPress plugin
...
       Success: Uninstalled 2 of 2 plugins.
       """
+
+  Scenario: Excluding a plugin from uninstallation when using --all switch
+    When I try `wp plugin uninstall --all --exclude=akismet,hello`
+    Then STDOUT should be:
+      """
+      Success: No plugins uninstalled.
+      """
+    And the return code should be 0
```
Comment (swissspidy) [inline features/plugin-uninstall.feature:81]: Nit: Missing newline at the end of the file

File: src/Plugin_Command.php [hunk 5]
```diff
@@ -375,6 +379,9 @@ public function activate( $args, $assoc_args = array() ) {
 	 * [--all]
 	 * : If set, all plugins will be deactivated.
 	 *
+	 *  [--exclude=<name>]
+	 * : Comma separated list of plugin names that should be excluded from deactivation.
```
Comment (swissspidy) [inline src/Plugin_Command.php:383]:
```suggestion
	 * : Comma separated list of plugin slugs that should be excluded from deactivation.
```

For consistency. It says slugs for the other two commands as well.

File: src/WP_CLI/ParsePluginNameInput.php [hunk 6]
```diff
@@ -13,16 +13,27 @@ trait ParsePluginNameInput {
...
 		}
 
+		if ( $all && $exclude ) {
+			$exclude_list = explode( ',', trim( $exclude, ',' ) );
+			foreach ( $exclude_list as $exclude_item ) {
+				$exclude_index = array_search( $exclude_item, $args, true );
+				if ( false !== $exclude_index ) {
+					unset( $args[ $exclude_index ] );
+				}
+			}
```
Comment (swissspidy) [inline src/WP_CLI/ParsePluginNameInput.php:34]: I think using `array_filter` improves readability here.

Something like:

```suggestion
			$args = array_filter(
				$args,
				static function( $slug ) {
					return ! in_array( $slug, $exclude_list );
				}
			);
```

File: features/plugin-uninstall.feature (hunk 4 above)
Comment (danielbachhuber) [reply features/plugin-uninstall.feature]: d03e01b7

File: features/plugin-delete.feature (hunk 3 above)
Comment (danielbachhuber) [reply features/plugin-delete.feature]: d03e01b7

File: features/plugin-activate.feature (hunk 1 above)
Comment (danielbachhuber) [reply features/plugin-activate.feature]: d03e01b7

File: src/WP_CLI/ParsePluginNameInput.php (hunk 6 above)
Comment (danielbachhuber) [reply src/WP_CLI/ParsePluginNameInput.php]: @swissspidy Updated with 62aec423

**Comment by wojsmol** [comment]
@danielbachhuber You can skip changes in `README.md` - it will be auto updated after this PR marge.

**Comment by danielbachhuber** [comment]
> You can skip changes in `README.md` - it will be auto updated after this PR marge.

@wojsmol Awesome, thanks!

**Review by swissspidy — APPROVED** [review]

**Pull request merged**